adam_preprocessing/shape_stroke_extraction.py

```diff
@@ -673,61 +689,68 @@ def strokes2controlpoints(self):
         Outputs the downsampled strokes to self.reduced_strokes. This should be an N x 10 x 2 array,
         where N is the number of strokes. The last dimension is the x vs y coords; the middle is the
         control point number.
+
+        Return:
+            reduced_strokes
         """
         reduced_strokes = []
         coef_inv = get_bsplines_matrix(10)
-        for i in range(len(self.strokes)):
-            reduced_strokes.append(reduced_stroke(self.strokes[i], coef_inv))
-        self.reduced_strokes = reduced_strokes = np.array(reduced_strokes, dtype=np.float64)
+        for i in range(len(strokes)):
+            reduced_strokes.append(reduced_stroke(strokes[i], coef_inv))
+        return np.array(reduced_strokes, dtype=np.float64)
 
-    def plot_strokes(self):
+    def plot_strokes(self, extractions: Sequence[ExtractionResult]):
         """
         Create stroke image, saving the result to self.stroke_img_save_path.
         """
-        stroke_colors = []
-        for j in range(self.num_obj):
-            reduced_obj = self.reduced_strokes[np.where(self.stroke_obj_ids == j)[0]]
-            plt.subplot(1, self.num_obj, j + 1)
-            s_c = []
-            for i in range(len(reduced_obj)):
-                p = plt.plot(reduced_obj[i][:, 1], reduced_obj[i][:, 0], linewidth=3)
-                s_c.append(p[0].get_color())
-                plt.scatter(reduced_obj[i][:, 1], reduced_obj[i][:, 0], c="r")
-            stroke_colors.append(s_c)
-            plt.gca().invert_yaxis()
-            plt.gca().set_aspect("equal", adjustable="box")
-            plt.axis("off")
+        stroke_colors_list = []
+        for extraction in extractions:
+            stroke_colors = []
+            for j in range(extraction.num_obj):
+                reduced_obj = extraction.reduced_strokes[np.where(extraction.stroke_obj_ids == j)[0]]
+                plt.subplot(1, extraction.num_obj, j + 1)
+                s_c = []
+                for i in range(len(reduced_obj)):
+                    p = plt.plot(reduced_obj[i][:, 1], reduced_obj[i][:, 0], linewidth=3)
+                    s_c.append(p[0].get_color())
+                    plt.scatter(reduced_obj[i][:, 1], reduced_obj[i][:, 0], c="r")
+                stroke_colors.append(s_c)
+                plt.gca().invert_yaxis()
+                plt.gca().set_aspect("equal", adjustable="box")
+                plt.axis("off")
+            stroke_colors_list.append(stroke_colors)
```

Comment:
I think there is something going wrong with this; the stroke output image (`stroke_0.png`) looks right when there are just a handful of masks per file, but it looks totally wrong when there are lots of masks in an image. I don't know what a lot of this pyplot stuff is doing but I'm wondering if some of these methods should be called only once. Also not sure how much effort this is worth fixing.

Reply:
Here's an example of the bad stroke plotting:

<details><summary>rgb_0.png</summary>

![rgb_0](https://user-images.githubusercontent.com/6520892/201834352-28331c41-81ce-4aad-924e-50d704a6b045.png)

</details>

<details><summary>semantic_0.png</summary>

![semantic_0](https://user-images.githubusercontent.com/6520892/201834359-5aac57c8-c450-4d5c-b432-eeaecd54dcc9.png)

</details>

<details><summary>stroke_0.png</summary>

![stroke_0](https://user-images.githubusercontent.com/6520892/201834423-c10d8961-7959-435c-a37f-5b8982dd1fc3.png)

</details>

<details><summary>stroke_graph_0.png</summary>

![stroke_graph_0](https://user-images.githubusercontent.com/6520892/201834425-509e229e-026c-40f0-a2bb-6bab830bb056.png)

Reply:
I've worked on this and here is the output for `situation_1`

<details><summary>rgb_0.png</summary>

![rgb_0](https://user-images.githubusercontent.com/6520892/202299409-cf99e8c6-997e-49ba-b47a-8358e26d4442.png)

</details>

<details><summary>semantic_0.png</summary>

![semantic_0](https://user-images.githubusercontent.com/6520892/202299411-c0659fdf-90c5-417f-9c67-ef3ae975a974.png)

</details>

<details><summary>stroke_0.png</summary>

![stroke_0](https://user-images.githubusercontent.com/6520892/202298412-53538d95-f125-4546-b60e-0a7e5f90378f.png)

</details>

<details><summary>stroke_graph_0.png</summary>

![stroke_graph_0](https://user-images.githubusercontent.com/6520892/202298417-0e7fabf4-11bb-4fac-917c-913c2490ff11.png)

</details>

These look right to me (minus the fact that the original extracted strokes are bad due to the background not being black). How does it look to you?

Reply:
Much better! Yes, this looks right to me.